[IR] improve hashing quality for ValueInfo

nocchijiang opened the pull request:
The current hashing quality for `ValueInfo` is poor because it uses pointers as the hash value, which can negatively impact performance in various places that use a `DenseSet`/`Map` of `ValueInfo`. In one observed case, `ModuleSummaryIndex::propagateAttributes()` was taking about 25 minutes to complete on a ThinLTO application. Profiling revealed that the majority of this time was spent operating on the `MarkedNonReadWriteOnly` set.

With the improved hashing, the execution time for `propagateAttributes` is dramatically reduced to less than 10 seconds.

teresajohnson reviewed (changes requested):
I don't like the approach of putting temporary analysis information in the GlobalValueSummaryInfo struct - it doesn't belong there logically, and while probably minimal, it also adds size overhead to the structure across all use cases.

I'm surprised to hear that this change took a 25 minute indexing down to 10s. We build very large apps frequently and I haven't seen anything like this, so I think we'll want to look more at why this is taking so long in your case. From the profile, is most of the time spent on the set insert, or on the set contains check?

It would be slightly better to put this in the Flags in the ValueInfo itself (where we already have the read only / write only flags being checked here), but I'd like a better understanding of why this is behaving so extreme in your app first.

teresajohnson commented on llvm/lib/IR/ModuleSummaryIndex.cpp:
```diff
@@ -213,9 +211,10 @@ propagateAttributesToRefs(GlobalValueSummary *S,
-      if (!MarkedNonReadWriteOnly.insert(VI).second)
+      if (VI.getRef()->second.MarkedNonReadWriteOnly)
         continue;
-    } else if (MarkedNonReadWriteOnly.contains(VI))
```
I think this check might be useless and possibly can be removed. We only get here and do the else if check if getAccessSpecifier() is non-zero, which means the VI is still either read or write only. But we only add a VI to the MarkedNonReadWriteOnly set if its getAccessSpecifier() is 0, and we only ever go towards 0 (marking things not read or write only), never the reverse.

teresajohnson replied:
I would add an assert in the else case though to confirm this assumption (that MarkedNonReadWriteOnly.contains(VI) is false).

teresajohnson reviewed (commented):
A couple ideas for improving the efficiency of this code below. I'm curious as to what the index graph looks like in your case. This isn't a recursive algorithm, so there presumably are just an extreme number of references for it to take so long. But then I'm surprised that improving this would drop the indexing time so dramatically to complete in 10s, as the rest of the indexing contains other graph traversals such as computing imports, and those aren't free.

nocchijiang commented:
Thanks for your quick feedback!

> I don't like the approach of putting temporary analysis information in the GlobalValueSummaryInfo struct - it doesn't belong there logically, and while probably minimal, it also adds size overhead to the structure across all use cases.

I don't like it neither. I will explain the reason below.

> I'm surprised to hear that this change took a 25 minute indexing down to 10s. We build very large apps frequently and I haven't seen anything like this, so I think we'll want to look more at why this is taking so long in your case. From the profile, is most of the time spent on the set insert, or on the set contains check?

During my investigation of an LTO miscompile affecting two iOS applications with predominantly Objective-C code, I encountered the performance issue being addressed.

The sample profile generated using `Instruments.app` from Xcode revealed only a single operation `grow` under  `propagateAttributes`. However, analyzing a few sample stack traces from `Console.app`, I observed that the leaf frames consistently manipulating the `MarkedNonReadWriteOnly` set. Also, I recorded the profile with a release version of `libLTO.dylib` as the debug version takes forever to complete. Therefore, it is challenging to pinpoint the specific operation that consumes more time.

I have noticed that you have made several suggestions in the review comments that I would definitely like to try and tell you how they would work.

> It would be slightly better to put this in the Flags in the ValueInfo itself (where we already have the read only / write only flags being checked here), but I'd like a better understanding of why this is behaving so extreme in your app first.

I must admit that I am not familiar with the code related to ThinLTO. The reason I eventually chose to put the flag in `GlobalValueSummaryInfo` is that `DenseMapInfo<ValueInfo>` utilizes the pointer identity of `GlobalValueSummaryMapTy` entry. If you believe it is acceptable to store the data directly within `ValueInfo`, I have no objections to that.

nocchijiang commented:
I tested the 2 ideas from the review comments. Unfortunately neither of them make noticeable improvements and I believe `DenseSet` overhead has to be avoided with a reasonable cost. Moving `MarkedNonReadWriteOnly` to `ValueInfo` and redesigning the inline flags to fit RO/WO/MarkedNonRWO into a 4-case enum results in a noticeable performance hit - now it takes 1 minute to complete `propagateAttributes()`. It appears that we must sacrifice some memory in order to enhance the performance. Thoughts?

nocchijiang commented:
I managed to grab some interesting profiling data with a RelWithDebInfo `libLTO.dylib`.

# Baseline

![image](https://github.com/user-attachments/assets/def0f19c-b6b0-43de-879a-b3febde4ded0)

Most time are spent in `insert`:

![image](https://github.com/user-attachments/assets/f3063d33-6ef7-4fe2-8cd8-98355fc741bb)

```
19.52 min   68.9%        0 s                             llvm::detail::DenseSetImpl<llvm::ValueInfo, llvm::DenseMap<llvm::ValueInfo, llvm::detail::DenseSetEmpty, llvm::DenseMapInfo<llvm::ValueInfo, void>, llvm::detail::DenseSetPair<llvm::ValueInfo>>, llvm::DenseMapInfo<llvm::ValueInfo, void>>::insert(llvm::ValueInfo const&)
1.15 min    4.0%        0 s                             llvm::detail::DenseSetImpl<llvm::ValueInfo, llvm::DenseMap<llvm::ValueInfo, llvm::detail::DenseSetEmpty, llvm::DenseMapInfo<llvm::ValueInfo, void>, llvm::detail::DenseSetPair<llvm::ValueInfo>>, llvm::DenseMapInfo<llvm::ValueInfo, void>>::contains(llvm::ValueInfo const&) const
```

# `MarkedNonReadWriteOnly` in `GlobalValueSummaryInfo`

![image](https://github.com/user-attachments/assets/06f5af9d-a73d-478c-9288-e34bb74d9400)

# `MarkedNonReadWriteOnly` in `ValueInfo`

![image](https://github.com/user-attachments/assets/4c20c3e2-7c65-44cd-9b49-53fd80e875d1)

![image](https://github.com/user-attachments/assets/c5a4ab1d-8fbf-4cb7-90ab-aec763c34188)

I strongly suspect that the relationship between `ValueInfo` and `GlobalValueSummaryInfo` is not strictly 1-to-1, otherwise the 10x performance difference cannot be explained. This is also the reason why I initially decided to declare the flag in `GlobalValueSummaryInfo`: `DenseMapInfo<ValueInfo>` utilizes the pointer identity of `GlobalValueSummaryMapTy` entry, so the `MarkedNonReadWriteOnly` set was effectively marking the underlying `GlobalValueSummaryInfo` rather than `ValueInfo`. However, I am uncertain whether this is the expected behavior.

nocchijiang commented:
The numbers are:
- set size: 23,373,881 at the end
- `propagateAttributesToRefs` is called 24,528,510 times
- 54,272,534 refs() get processed in that loop
- 48,428,967 of these have `!VI.getAccessSpecifier()`

nikic commented:
Inserting 50M values into a hash table shouldn't really take an appreciable amount of time. I wonder whether you're hitting some kind of catastrophic hash collision issue and complexity degenerates quadratically?

The implementation is
```
static unsigned getHashValue(ValueInfo I) { return (uintptr_t)I.getRef(); }
```
Which is already somewhat bad because the low three bits of I.getRef() are guaranteed to be zero, so you could try whether adding a `>> 3` improves things a bit. And then maybe try whether wrapping it in a hash_value() helps even more.

nocchijiang commented:
Applying the shift does help, but it is still far from acceptable.

```
Weight	Self Weight		Symbol Name
9.94 min   56.3%	929.00 ms	 	llvm::ModuleSummaryIndex::propagateAttributes(llvm::DenseSet<unsigned long long, llvm::DenseMapInfo<unsigned long long, void>> const&)
9.79 min   55.5%	1.96 s	 	 propagateAttributesToRefs(llvm::GlobalValueSummary*, llvm::DenseSet<llvm::ValueInfo, llvm::DenseMapInfo<llvm::ValueInfo, void>>&)
8.93 min   50.6%	0 s	 	  llvm::detail::DenseSetImpl<llvm::ValueInfo, llvm::DenseMap<llvm::ValueInfo, llvm::detail::DenseSetEmpty, llvm::DenseMapInfo<llvm::ValueInfo, void>, llvm::detail::DenseSetPair<llvm::ValueInfo>>, llvm::DenseMapInfo<llvm::ValueInfo, void>>::insert(llvm::ValueInfo const&)
8.93 min   50.6%	0 s	 	   std::__1::pair<llvm::DenseMapIterator<llvm::ValueInfo, llvm::detail::DenseSetEmpty, llvm::DenseMapInfo<llvm::ValueInfo, void>, llvm::detail::DenseSetPair<llvm::ValueInfo>, false>, bool> llvm::DenseMapBase<llvm::DenseMap<llvm::ValueInfo, llvm::detail::DenseSetEmpty, llvm::DenseMapInfo<llvm::ValueInfo, void>, llvm::detail::DenseSetPair<llvm::ValueInfo>>, llvm::ValueInfo, llvm::detail::DenseSetEmpty, llvm::DenseMapInfo<llvm::ValueInfo, void>, llvm::detail::DenseSetPair<llvm::ValueInfo>>::try_emplace<llvm::detail::DenseSetEmpty&>(llvm::ValueInfo const&, llvm::detail::DenseSetEmpty&)
8.93 min   50.6%	0 s	 	    bool llvm::DenseMapBase<llvm::DenseMap<llvm::ValueInfo, llvm::detail::DenseSetEmpty, llvm::DenseMapInfo<llvm::ValueInfo, void>, llvm::detail::DenseSetPair<llvm::ValueInfo>>, llvm::ValueInfo, llvm::detail::DenseSetEmpty, llvm::DenseMapInfo<llvm::ValueInfo, void>, llvm::detail::DenseSetPair<llvm::ValueInfo>>::LookupBucketFor<llvm::ValueInfo>(llvm::ValueInfo const&, llvm::detail::DenseSetPair<llvm::ValueInfo>*&)
8.93 min   50.6%	8.93 min	 	     bool llvm::DenseMapBase<llvm::DenseMap<llvm::ValueInfo, llvm::detail::DenseSetEmpty, llvm::DenseMapInfo<llvm::ValueInfo, void>, llvm::detail::DenseSetPair<llvm::ValueInfo>>, llvm::ValueInfo, llvm::detail::DenseSetEmpty, llvm::DenseMapInfo<llvm::ValueInfo, void>, llvm::detail::DenseSetPair<llvm::ValueInfo>>::LookupBucketFor<llvm::ValueInfo>(llvm::ValueInfo const&, llvm::detail::DenseSetPair<llvm::ValueInfo> const*&) const
```

Wrapping the shifted value in `hash_value()` does the trick:

```
Weight	Self Weight		Symbol Name
5.13 s    1.1%	577.00 ms	 	llvm::ModuleSummaryIndex::propagateAttributes(llvm::DenseSet<unsigned long long, llvm::DenseMapInfo<unsigned long long, void>> const&)
2.54 s    0.5%	1.57 s	 	 propagateAttributesToRefs(llvm::GlobalValueSummary*, llvm::DenseSet<llvm::ValueInfo, llvm::DenseMapInfo<llvm::ValueInfo, void>>&)
456.00 ms    0.1%	0 s	 	  llvm::detail::DenseSetImpl<llvm::ValueInfo, llvm::DenseMap<llvm::ValueInfo, llvm::detail::DenseSetEmpty, llvm::DenseMapInfo<llvm::ValueInfo, void>, llvm::detail::DenseSetPair<llvm::ValueInfo>>, llvm::DenseMapInfo<llvm::ValueInfo, void>>::contains(llvm::ValueInfo const&) const
456.00 ms    0.1%	50.00 ms	 	   llvm::DenseMapBase<llvm::DenseMap<llvm::ValueInfo, llvm::detail::DenseSetEmpty, llvm::DenseMapInfo<llvm::ValueInfo, void>, llvm::detail::DenseSetPair<llvm::ValueInfo>>, llvm::ValueInfo, llvm::detail::DenseSetEmpty, llvm::DenseMapInfo<llvm::ValueInfo, void>, llvm::detail::DenseSetPair<llvm::ValueInfo>>::find(llvm::ValueInfo const&) const
406.00 ms    0.0%	393.00 ms	 	    bool llvm::DenseMapBase<llvm::DenseMap<llvm::ValueInfo, llvm::detail::DenseSetEmpty, llvm::DenseMapInfo<llvm::ValueInfo, void>, llvm::detail::DenseSetPair<llvm::ValueInfo>>, llvm::ValueInfo, llvm::detail::DenseSetEmpty, llvm::DenseMapInfo<llvm::ValueInfo, void>, llvm::detail::DenseSetPair<llvm::ValueInfo>>::LookupBucketFor<llvm::ValueInfo>(llvm::ValueInfo const&, llvm::detail::DenseSetPair<llvm::ValueInfo> const*&) const
12.00 ms    0.0%	2.00 ms	 	     llvm::DenseMapBase<llvm::DenseMap<llvm::ValueInfo, llvm::detail::DenseSetEmpty, llvm::DenseMapInfo<llvm::ValueInfo, void>, llvm::detail::DenseSetPair<llvm::ValueInfo>>, llvm::ValueInfo, llvm::detail::DenseSetEmpty, llvm::DenseMapInfo<llvm::ValueInfo, void>, llvm::detail::DenseSetPair<llvm::ValueInfo>>::getHashValue(llvm::ValueInfo const&)
10.00 ms    0.0%	0 s	 	      llvm::DenseMapInfo<llvm::ValueInfo, void>::getHashValue(llvm::ValueInfo)
10.00 ms    0.0%	0 s	 	       std::__1::enable_if<is_integral_or_enum<unsigned long>::value, llvm::hash_code>::type llvm::hash_value<unsigned long>(unsigned long)
```

I will update the patch as per @nikic 's great suggestion.

nocchijiang committed:
[IR] improve hashing quality for ValueInfo

The current hashing quality for ValueInfo is poor because it uses
pointers as the hash value, which can negatively impact performance in
various places that use a DenseSet/Map of ValueInfo. In one observed
case, ModuleSummaryIndex::propagateAttributes() was taking about 25
minutes to complete on a ThinLTO application. Profiling revealed that
the majority of this time was spent operating on the
MarkedNonReadWriteOnly set.

With the improved hashing, the execution time for propagateAttributes is
dramatically reduced to less than 10 seconds.

nikic commented on llvm/include/llvm/IR/ModuleSummaryIndex.h:
```diff
@@ -300,7 +301,9 @@ template <> struct DenseMapInfo<ValueInfo> {
   }
-  static unsigned getHashValue(ValueInfo I) { return (uintptr_t)I.getRef(); }
+  static unsigned getHashValue(ValueInfo I) {
+    return hash_value(((uintptr_t)I.getRef() >> ValueInfo::NumFlagsBit));
```
There shouldn't be a need to use the shift if you're using hash_value. The shift helps because the low bits of the hash get used, but if you're using hash_value, it makes no different whether the low or high bits are zero, everything will get thoroughly mixed.

nocchijiang replied:
OK, I dropped the shift in the latest patch.

teresajohnson reviewed (approved):
@nikic nice find!

This could presumably help other places that use a DenseSet/Map of ValueInfo. Perhaps the description should be about improving the hashing of ValueInfo and note the massive improvement to propagateAttributes in your case in the summary.

teresajohnson commented:
> Updated the title and description of the PR and commit message as suggested by @teresajohnson. Could you also help me merge this PR since I don't have commit access?

Done!
